python/pyspark/worker.py

```diff
@@ -111,9 +114,16 @@ def wrapped(key_series, value_series):
                 "Number of columns of the returned pandas.DataFrame "
                 "doesn't match specified schema. "
                 "Expected: {} Actual: {}".format(len(return_type), len(result.columns)))
-        arrow_return_types = (to_arrow_type(field.dataType) for field in return_type)
-        return [(result[result.columns[i]], arrow_type)
-                for i, arrow_type in enumerate(arrow_return_types)]
+        try:
+            # Assign result columns by schema name
+            return [(result[field.name], to_arrow_type(field.dataType)) for field in return_type]
+        except KeyError:
+            if all(not isinstance(name, basestring) for name in result.columns):
+                # Assign result columns by position if they are not named with strings
+                return [(result[result.columns[i]], to_arrow_type(field.dataType))
+                        for i, field in enumerate(return_type)]
+            else:
+                raise
```

Comment:
Why we limit to just result columns not named with strings?

In the case we return a pd.DataFrame with matching field types, but not matching field names, we don't like to allow it?

If returned pd.DataFrame doesn't match return_type's column names, shouldn't we follow current behavior?

Reply:
I think when user specify column names explicitly on the returned pd.DataFrame but it doesn't match the schema, then it's most likely to be a bug / typo, so throw exception makes sense to me.

Reply:
Ah, I saw you add document for this behavior. Looks good.